persistent effects

Be-ing opened the pull request:
Save state of effect chains to `effects.xml` in user settings directory on shutdown and reload state of effects chains from that file on startup. This paves the way for exporting & importing effect chain presets.

Should the state of effect metaknobs and/or chain superknobs be saved as well?

Be-ing committed:
implement effect saving to/loading from XML

Be-ing committed:
save/load effect parameter state to XML on shutdown/startup

sblaisot commented:
Can you please provide the XML schema used? How is it upward and backward compatible between mixxx versions?

Be-ing commented:
I have not written a formal schema, but here is an example `effects.xml` file:
```
<?xml version='1.0' encoding='utf-8'?>
<MixxxEffects>
 <Rack>
  <Group>[EffectRack1]</Group>
  <Chains>
   <EffectChain>
    <Name></Name>
    <Description></Description>
    <InsertionType>0</InsertionType>
    <Effects>
     <Effect>
      <Id>org.mixxx.effects.filter</Id>
      <Version>1.0</Version>
      <Parameters>
       <KnobParameters>
        <KnobParameter>
         <Id>lpf</Id>
         <Value>1</Value>
         <LinkType>2</LinkType>
         <LinkInversion>1</LinkInversion>
        </KnobParameter>
        <KnobParameter>
         <Id>q</Id>
         <Value>0.487611</Value>
         <LinkType>0</LinkType>
         <LinkInversion>0</LinkInversion>
        </KnobParameter>
        <KnobParameter>
         <Id>hpf</Id>
         <Value>0.322835</Value>
         <LinkType>3</LinkType>
         <LinkInversion>0</LinkInversion>
        </KnobParameter>
       </KnobParameters>
       <ButtonParameters/>
      </Parameters>
     </Effect>
     <Effect/>
     <Effect>
      <Id>org.mixxx.effects.phaser</Id>
      <Version>1.0</Version>
      <Parameters>
       <KnobParameters>
        <KnobParameter>
         <Id>depth</Id>
         <Value>0</Value>
         <LinkType>1</LinkType>
         <LinkInversion>0</LinkInversion>
        </KnobParameter>
        <KnobParameter>
         <Id>range</Id>
         <Value>0.645669</Value>
         <LinkType>1</LinkType>
         <LinkInversion>0</LinkInversion>
        </KnobParameter>
        <KnobParameter>
         <Id>feedback</Id>
         <Value>0.716535</Value>
         <LinkType>0</LinkType>
         <LinkInversion>0</LinkInversion>
        </KnobParameter>
        <KnobParameter>
         <Id>stages</Id>
         <Value>0.311024</Value>
         <LinkType>0</LinkType>
         <LinkInversion>0</LinkInversion>
        </KnobParameter>
        <KnobParameter>
         <Id>lfo_frequency</Id>
         <Value>0.399213</Value>
         <LinkType>0</LinkType>
         <LinkInversion>0</LinkInversion>
        </KnobParameter>
       </KnobParameters>
       <ButtonParameters>
        <ButtonParameter>
         <Id>stereo</Id>
         <Value>1</Value>
        </ButtonParameter>
       </ButtonParameters>
      </Parameters>
     </Effect>
     <Effect/>
    </Effects>
   </EffectChain>
   <EffectChain>
    <Name></Name>
    <Description></Description>
    <InsertionType>0</InsertionType>
    <Effects>
     <Effect/>
     <Effect>
      <Id>org.mixxx.effects.autopan</Id>
      <Version>1.0</Version>
      <Parameters>
       <KnobParameters>
        <KnobParameter>
         <Id>width</Id>
         <Value>0</Value>
         <LinkType>1</LinkType>
         <LinkInversion>0</LinkInversion>
        </KnobParameter>
        <KnobParameter>
         <Id>period</Id>
         <Value>1</Value>
         <LinkType>1</LinkType>
         <LinkInversion>1</LinkInversion>
        </KnobParameter>
        <KnobParameter>
         <Id>smoothing</Id>
         <Value>0.851646</Value>
         <LinkType>0</LinkType>
         <LinkInversion>0</LinkInversion>
        </KnobParameter>
       </KnobParameters>
       <ButtonParameters>
        <ButtonParameter>
         <Id>periodUnit</Id>
         <Value>1</Value>
        </ButtonParameter>
       </ButtonParameters>
      </Parameters>
     </Effect>
     <Effect/>
     <Effect/>
    </Effects>
   </EffectChain>
   <EffectChain>
    <Name></Name>
    <Description></Description>
    <InsertionType>0</InsertionType>
    <Effects>
     <Effect/>
     <Effect/>
     <Effect/>
     <Effect/>
    </Effects>
   </EffectChain>
   <EffectChain>
    <Name></Name>
    <Description></Description>
    <InsertionType>0</InsertionType>
    <Effects>
     <Effect/>
     <Effect/>
     <Effect/>
     <Effect/>
    </Effects>
   </EffectChain>
  </Chains>
 </Rack>
</MixxxEffects>
```

For backwards compatibility, if `effects.xml` is unreadable (or is not valid XML), blank effect chains are loaded like before. Also, if any element is empty or missing, a blank chain or effect or the default state of the parameter is loaded. Empty `<Effect/>` elements are intentionally used to keep the order of chains. Extra elements are ignored.

For forwards compatibility, it would be a good idea to indicate the Mixxx version with the `<MixxxEffects>` element, so I'll add that. I talked with @rryan on IRC and we agreed that a good way to handle storage of saved/imported effect chain presets in the future would be a similar XML file. I think it would be good to structure that like the `<Chains>` element above, with an arbitrary number of `<EffectChain>` child elements. Those would be saved in a different XML file, so case something goes wrong in `effects.xml` users won't lose their imported chains. I initially was thinking about saving one effect chain per file and relying on the file name to name them, but @rryan pointed out this would not be a good idea because of incompatibilities between filesystems and their limitations on valid file names. There is already a `<Name>` child element of `<EffectChain>`, but there is no UI to edit the names of chains currently this will be empty. If any other effect parameter types are added in the future (for example enum parameters), it would be easy to add another child element of `<Parameters>`.

Be-ing committed:
add Mixxx version to effects.xml

Be-ing committed:
make effect parameters loaded from XML actually change the sound

Be-ing committed:
organize EffectParameterSlot contructor

Be-ing commented on src/effects/effectparameterslot.cpp:
```diff
@@ -243,3 +246,42 @@ void EffectParameterSlot::slotValueChanged(double v) {
+    } else {
+        m_pControlValue->setParameter(
+            XmlParse::selectNodeDouble(knobParameterElement, "Value"));
+        m_pEffectParameter->setValue(m_pControlValue->get());
```
I don't understand why this is necessary. m_pControlValue's valueChanged signal is connected to EffectParameterSlot::slotValueChanged in the EffectParameterSlot constructor, so the line above should already do this.

daschuer replied:
setParameter does not send value change event to the owner. Please try setParamerterFrom( .. , nullptr); 
I think you have also to deal with soft takeover since controllers will be out of sync:
m_pSoftTakeover->ignore(m_pControlValue, parameter);

Be-ing replied:
I tried `m_pControlValue->set()` and it had the same issue.

Soft takeover should not be active here. That would interfere with receiving initial positions of knobs from controllers. I have designed the Controls JS library to accommodate that.

Be-ing replied:
`setParameterFrom(..., nullptr)` does work. I don't understand how. Could you explain? I looked through the ControlObject and ControlDoublePrivate source but it isn't clear to me. I see that ControlDoublePrivate::set() has the line `emit(valueChanged(value, pSender));`, but I don't understand what happens when the pSender argument is omitted.

Be-ing committed:
small code cleanup

daschuer commented:
Saving the Mixxx version in the xml file will force us to migrate the file on every Mixxx verison change. 
It also does not allow us to change the xml schema during development without changing Mixxx version. 
I think we should version the schema itself and remove the Mixxx version.

daschuer commented on src/effects/effectchainmanager.cpp:
```diff
@@ -164,37 +168,58 @@ bool EffectChainManager::saveEffectChains() {
-        return false;
+        EffectChainPointer pEmptyChain;
+        QDomElement emptyChainElement = doc.createElement("EffectChain");
+        for (int i = 0; i < 4; ++i) {
```
the 4 should be a common definition

Be-ing replied:
I agree. Where?

daschuer replied:
Is this the same 4 we have in StandardEffectRack::addEffectChainSlot? 
Maybe it can become a const member of EffectChainManager,

daschuer commented on src/effects/effectchainslot.cpp:
```diff
@@ -390,3 +391,51 @@ void EffectChainSlot::slotChannelStatusChanged(const QString& group) {
+        return chainElement;
+    }
+
+    XmlParse::addElement(*doc, chainElement, "Name",
```
Strings like "Name" should be placed in an anonymous namespace on top of this file.

Be-ing replied:
Anonymous namespaces will not work well for this because these strings are reused in both the Slot (toXML & loadValuesFromXml) and non-Slot (createFromXml) classes in src/effects. Can you suggest a better place to define the strings?

daschuer replied:
Than you might use an independent header file that is included by both

Be-ing committed:
define a constant for the number of effects

Be-ing committed:
version effects.xml on its own terms instead of by Mixxx version

daschuer commented on src/effects/effectchainmanager.h:
```diff
@@ -53,6 +53,8 @@ class EffectChainManager : public QObject {
     bool saveEffectChains();
     QList<std::pair<EffectChainPointer, QDomElement>> loadEffectChains();
 
+    static const int kNumEffectsPerUnit = 4;
```
I am not sure if this works nice on all compilers. 
I would prefer to wrap this into a getter and move this constant to the cpp file

daschuer commented on src/effects/effectsmanager.h:
```diff
@@ -22,6 +22,10 @@
 
+// Version history:
+// 0 (Mixxx 2.1.0): initial support for saving state of effects
+const int EFFECT_XML_VERSION = 0;
```
kEffectXmlVersion

MIXXX_VERSION is a macro

Be-ing committed:
fix segfault when switching effect chains

Be-ing committed:
move effect XML element names to separate header file

Be-ing committed:
save superknob value to XML

Be-ing committed:
save metaknob value to XML

Be-ing commented:
@daschuer any more issues?

Be-ing commented:
Whoops, I forgot to add the new header file in 61669a5. Added it in a new commit.

Be-ing commented:
Let's not merge this until #1118 is merged to avoid having to increase the effect version numbers.

Be-ing commented:
Okay, all merge conflicts and build issues fixed. Tests pass again. Ready for review.

Be-ing committed:
fix effects tests

Be-ing commented:
@daschuer any more comments? Ready to merge?

Be-ing commented:
I'd like to get this merged soon.

rryan commented:
I definitely want us to be doing this, but I'm feeling nervous about putting this into 2.1 at the last minute. I think it's worth getting some testing in master before we let it get out to users. What do you think about delaying this to 2.2?

I also think we should probably get the effect versioning story figured out (e.g. what to do when we load an effect.xml which has an out-dated version for an effect stored or settings for a newer version). Native effects should probably implement an upgrade method that knows how to migrate old parameters to new ones.

Be-ing commented:
I think this should be in 2.1. Without this, setting up custom effect chains is impractical because they'd have to be configured from scratch each time Mixxx is started. That greatly limits the usefulness of the effects system.

We should have at least a vague idea for how to deal with effect version upgrades, but we don't need to figure out everything until we have effects to upgrade.

Be-ing commented:
I'm a bit confused why there are effect version numbers. How might an effect change in the future such that restoring its saved state would need upgrade code? New default parameter states wouldn't warrant a version bump. If parameters are rearranged, the state loading code could be modified so the old parameter states are applied to the correct parameters.

rryan commented:
> I'm a bit confused why there are effect version numbers. How might an effect change in the future such that restoring its saved state would need upgrade code? 

I think the recent reverb changes are an example -- we changed the meaning of a parameter. 

If we change the interpretation of a parameter -- for example changing from log to linear, we could re-map the stored value appropriately to fit the new curve.

> New default parameter states wouldn't warrant a version bump.

Yes -- though the saved values could lie outside the effect min/max values.

> If parameters are rearranged, the state loading code could be modified so the old parameter states are applied to the correct parameters.

Parameter IDs let us restore based on ID rather than order, so order is free to change. (Or if we ever allow user's to re-order parameters in Preferences -> Effects or something).

daschuer commented:
How do LV2 effects fit to this concept?

rryan commented:
> How do LV2 effects fit to this concept?

Do you mean RE: versioning or the persistence in general?

daschuer commented:
In this case I mean versioning, because the LV2 effect versions are independent from Mixxx.  Integrate them in general should be IMHO no big issue, once we have a solution for the n LV2 parameter to 8 Mixxx parameter issue.

daschuer commented:
By the way, is your mixxx.org email alive?

rryan commented:
> In this case I mean versioning, because the LV2 effect versions are independent from Mixxx. Integrate them in general should be IMHO no big issue, once we have a solution for the n LV2 parameter to 8 Mixxx parameter issue.

Right -- for LV2 we can't really do anything like we're talking about for native effects. We will already be recording parameters by ID so we can match them up properly if the LV2 effect re-orders them and do some bounds checking on the stored values relative to the latest manifest loaded from the plugin.

rryan commented:
Could you please remove the distinction between ButtonParameter and KnobParameter in the XML (i.e. make them all simply Parameter). That seems to leak an artifact of our implementation into the file format.

Be-ing committed:
save all parameter types as a generic <Parameter> element in XML

Be-ing commented:
> Could you please remove the distinction between ButtonParameter and KnobParameter in the XML (i.e. make them all simply Parameter). That seems to leak an artifact of our implementation into the file format.

Coincidentally, to implement this I had to make restoring saved state independent of the parameter order in XML.

rryan commented:
> Coincidentally, to implement this I made restoring saved state independent of the parameter order in XML.

Great :)

Be-ing commented:
Travis passed, Appveyor is stalled.

rryan commented:
Hm, we should probably use enum names instead of integer values in the XML so we aren't tied to keeping the integer values the same.

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into effect_xml

Be-ing committed:
save metaknob link type as a string in XML

Be-ing commented:
> Hm, we should probably use enum names instead of integer values in the XML so we aren't tied to keeping the integer values the same.

I did this for the link type. I don't think it's necessary for the link inversion boolean value. Any other issues?

Be-ing commented:
The chain insertion type should probably also be saved as a string. I'll take care of that.

Be-ing committed:
save effect chain insertion type as a string in XML

Be-ing commented:
Here is how effects.xml looks now:
```
<?xml version='1.0' encoding='utf-8'?>
<MixxxEffects>
 <SchemaVersion>0</SchemaVersion>
 <Rack>
  <Group>[EffectRack1]</Group>
  <Chains>
   <EffectChain>
    <Id></Id>
    <Name></Name>
    <Description></Description>
    <InsertionType>INSERT</InsertionType>
    <SuperParameterValue>1</SuperParameterValue>
    <Effects>
     <Effect/>
     <Effect>
      <MetaParameterValue>0.677165</MetaParameterValue>
      <Id>org.mixxx.effects.linkwitzrileyeq</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>high</Id>
        <Value>0.5</Value>
        <LinkType>LINK_NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>killHigh</Id>
        <Value>0</Value>
       </Parameter>
       <Parameter>
        <Id>killLow</Id>
        <Value>0</Value>
       </Parameter>
       <Parameter>
        <Id>killMid</Id>
        <Value>0</Value>
       </Parameter>
       <Parameter>
        <Id>low</Id>
        <Value>0.5</Value>
        <LinkType>LINK_NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>mid</Id>
        <Value>0.5</Value>
        <LinkType>LINK_NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect>
      <MetaParameterValue>0.84252</MetaParameterValue>
      <Id>org.mixxx.effects.reverb</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>bandwidth</Id>
        <Value>0.15748</Value>
        <LinkType>LINKED</LinkType>
        <LinkInversion>1</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>damping</Id>
        <Value>0</Value>
        <LinkType>LINK_NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>decay</Id>
        <Value>0.5</Value>
        <LinkType>LINK_NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>send_amount</Id>
        <Value>0.84252</Value>
        <LinkType>LINKED</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect/>
    </Effects>
   </EffectChain>
   <EffectChain>
    <Id></Id>
    <Name></Name>
    <Description></Description>
    <InsertionType>INSERT</InsertionType>
    <SuperParameterValue>0</SuperParameterValue>
    <Effects>
     <Effect>
      <MetaParameterValue>0.165354</MetaParameterValue>
      <Id>org.mixxx.effects.autopan</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>period</Id>
        <Value>0.256306</Value>
        <LinkType>LINK_NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>periodUnit</Id>
        <Value>1</Value>
       </Parameter>
       <Parameter>
        <Id>smoothing</Id>
        <Value>1</Value>
        <LinkType>LINK_NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>width</Id>
        <Value>0.165354</Value>
        <LinkType>LINKED</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect>
      <MetaParameterValue>0.598425</MetaParameterValue>
      <Id>org.mixxx.effects.moogladder4filter</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>hpf</Id>
        <Value>0.19685</Value>
        <LinkType>LINKED_RIGHT</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>lpf</Id>
        <Value>1</Value>
        <LinkType>LINKED_LEFT</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>resonance</Id>
        <Value>0.5</Value>
        <LinkType>LINK_NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect>
      <MetaParameterValue>0.393701</MetaParameterValue>
      <Id>org.mixxx.effects.autopan</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>period</Id>
        <Value>0.212598</Value>
        <LinkType>LINKED_LEFT_RIGHT</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>periodUnit</Id>
        <Value>1</Value>
       </Parameter>
       <Parameter>
        <Id>smoothing</Id>
        <Value>1</Value>
        <LinkType>LINK_NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>width</Id>
        <Value>0.393701</Value>
        <LinkType>LINKED</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect/>
    </Effects>
   </EffectChain>
   <EffectChain>
    <Id></Id>
    <Name></Name>
    <Description></Description>
    <InsertionType>INSERT</InsertionType>
    <SuperParameterValue>0</SuperParameterValue>
    <Effects>
     <Effect/>
     <Effect/>
     <Effect/>
     <Effect/>
    </Effects>
   </EffectChain>
   <EffectChain>
    <Id></Id>
    <Name></Name>
    <Description></Description>
    <InsertionType>INSERT</InsertionType>
    <SuperParameterValue>0</SuperParameterValue>
    <Effects>
     <Effect/>
     <Effect/>
     <Effect/>
     <Effect/>
    </Effects>
   </EffectChain>
  </Chains>
 </Rack>
</MixxxEffects>
```

Nothing we might regret doing in the format jumps out to me at this point.

Be-ing committed:
make strings for LinkType match enum names used for LinkType

Be-ing commented:
I'm a bit torn whether to remove the redundant "LINK_" prefix for the strings in the XML file or make those strings match the enum names. It doesn't matter too much either way, but we should pick one.

rryan commented:
> Nothing we might regret doing in the format jumps out to me at this point.

Cool -- thanks for the changes! I tried it out and it seems to work as advertised. 

To match the controller root document node:
```xml
<MixxxMIDIPreset schemaVersion="1" mixxxVersion="1.9.0+">
```
should schemaVersion be an attribute of the ```MixxxEffects``` tag?

rryan commented:
> I'm a bit torn whether to remove the redundant "LINK_" prefix for the strings in the XML file or make those strings match the enum names. It doesn't matter too much either way, but we should pick one.

The LINK_ prefix is only needed in C++ to add some fake scoping since it's an old-style enum (not a c++11 enum class) which pollutes the scope you define it in. We could add enum class and remove the LINK_ prefixes in the code. I think you don't really need it in the XML since there's no ambiguity (since it's inside a LinkType node).

Be-ing committed:
convert EffectManifestParameter LinkType to enum class

Be-ing committed:
save effect XML schema version as attribute of root element

Be-ing commented:
Okay, here is what the XML file looks like now:
```
<?xml version='1.0' encoding='utf-8'?>
<MixxxEffects schemaVersion="0">
 <Rack>
  <Group>[EffectRack1]</Group>
  <Chains>
   <EffectChain>
    <Id></Id>
    <Name></Name>
    <Description></Description>
    <InsertionType>INSERT</InsertionType>
    <SuperParameterValue>1</SuperParameterValue>
    <Effects>
     <Effect>
      <MetaParameterValue>0</MetaParameterValue>
      <Id>org.mixxx.effects.filter</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>hpf</Id>
        <Value>0</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>lpf</Id>
        <Value>1</Value>
        <LinkType>LINKED</LinkType>
        <LinkInversion>1</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>q</Id>
        <Value>0.487611</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect>
      <MetaParameterValue>0</MetaParameterValue>
      <Id>org.mixxx.effects.linkwitzrileyeq</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>high</Id>
        <Value>0.5</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>killHigh</Id>
        <Value>0</Value>
       </Parameter>
       <Parameter>
        <Id>killLow</Id>
        <Value>0</Value>
       </Parameter>
       <Parameter>
        <Id>killMid</Id>
        <Value>0</Value>
       </Parameter>
       <Parameter>
        <Id>low</Id>
        <Value>0.5</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>mid</Id>
        <Value>0.5</Value>
        <LinkType>LINKED</LinkType>
        <LinkInversion>1</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect>
      <MetaParameterValue>0.551181</MetaParameterValue>
      <Id>org.mixxx.effects.reverb</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>bandwidth</Id>
        <Value>0.551181</Value>
        <LinkType>LINKED</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>damping</Id>
        <Value>0</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>decay</Id>
        <Value>0.102362</Value>
        <LinkType>LINKED_LEFT_RIGHT</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>send_amount</Id>
        <Value>0</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect/>
    </Effects>
   </EffectChain>
   <EffectChain>
    <Id></Id>
    <Name></Name>
    <Description></Description>
    <InsertionType>INSERT</InsertionType>
    <SuperParameterValue>0</SuperParameterValue>
    <Effects>
     <Effect>
      <MetaParameterValue>0.740157</MetaParameterValue>
      <Id>org.mixxx.effects.moogladder4filter</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>hpf</Id>
        <Value>0.480315</Value>
        <LinkType>LINKED_RIGHT</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>lpf</Id>
        <Value>1</Value>
        <LinkType>LINKED_LEFT</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>resonance</Id>
        <Value>0.480315</Value>
        <LinkType>LINKED_LEFT_RIGHT</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect>
      <MetaParameterValue>0.598425</MetaParameterValue>
      <Id>org.mixxx.effects.moogladder4filter</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>hpf</Id>
        <Value>0</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>lpf</Id>
        <Value>1</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>resonance</Id>
        <Value>0.5</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect>
      <MetaParameterValue>0.393701</MetaParameterValue>
      <Id>org.mixxx.effects.autopan</Id>
      <Version>1.0</Version>
      <Parameters>
       <Parameter>
        <Id>period</Id>
        <Value>0.256306</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>periodUnit</Id>
        <Value>1</Value>
       </Parameter>
       <Parameter>
        <Id>smoothing</Id>
        <Value>1</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
       <Parameter>
        <Id>width</Id>
        <Value>0.5</Value>
        <LinkType>NONE</LinkType>
        <LinkInversion>0</LinkInversion>
       </Parameter>
      </Parameters>
     </Effect>
     <Effect/>
    </Effects>
   </EffectChain>
   <EffectChain>
    <Id></Id>
    <Name></Name>
    <Description></Description>
    <InsertionType>INSERT</InsertionType>
    <SuperParameterValue>0</SuperParameterValue>
    <Effects>
     <Effect/>
     <Effect/>
     <Effect/>
     <Effect/>
    </Effects>
   </EffectChain>
   <EffectChain>
    <Id></Id>
    <Name></Name>
    <Description></Description>
    <InsertionType>INSERT</InsertionType>
    <SuperParameterValue>0</SuperParameterValue>
    <Effects>
     <Effect/>
     <Effect/>
     <Effect/>
     <Effect/>
    </Effects>
   </EffectChain>
  </Chains>
 </Rack>
</MixxxEffects>
```

Be-ing committed:
replace remaining enums in EffectManifestParameter with enum classes

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into effect_xml

use new enum classes for effect parameter manifests

Be-ing committed:
fix metaknob_link_test.cpp build

Be-ing commented:
Travis passed.

daschuer commented:
Do we need 
<SuperParameterValue>1</SuperParameterValue> and 
<MetaParameterValue>0.551181</MetaParameterValue>
? 
Following our A/B discussion https://github.com/mixxxdj/mixxx/pull/1148, I think it is better to not have them here. This would add an other degree of freedom/confusion, because they may conflicts with the stored parameter set. 

Once we have a preference option that fixes https://github.com/mixxxdj/mixxx/pull/1148, the Super and Meta should either be unchanged and overwrite the stored parameter values, to keep the controller in sync (A), or the super / meta knob should adopt the effect defaults value, to archive the stored sound but the fixed scaled controller knobs out of sync (B).

daschuer commented on src/effects/effectbuttonparameterslot.cpp:
```diff
@@ -107,3 +109,33 @@ void EffectButtonParameterSlot::slotValueChanged(double v) {
+        XmlParse::addElement(*doc, parameterElement,
+                             EffectXml::ParameterId,
+                             m_pEffectParameter->id());
+        // TODO(rryan): Do smarter QVariant formatting?
```
This TODO seams to be obsolete. addElement() requires QString and QString::number suits for that.

daschuer commented on src/effects/effectslot.cpp:
```diff
@@ -268,3 +272,58 @@ void EffectSlot::slotEffectMetaParameter(double v, bool force) {
+    return effectElement;
+}
+
+void EffectSlot::loadValuesFromXml(const QDomElement& effectElement) {
```
We have some loadValuesFromXml() function all do different things. Can we name this accordingly? 
This could become loadParametersFromXml() 
And below: 
pParameterSlot->loadValueFromXml()

daschuer replied:
or pParameterSlot->setupParameterFromXml() and this setupParametersFromXml()

Be-ing commented:
>Do we need <SuperParameterValue>1</SuperParameterValue> and <MetaParameterValue>0.551181</MetaParameterValue>?

I don't understand why we would store the parameter values but not the metaknob and superknob values. It's possible the user intentionally set parameters out of sync with the metaknob and superknob. I don't think Mixxx should automatically override that.

daschuer commented on src/effects/effectparameterslot.cpp:
```diff
@@ -242,3 +253,54 @@ void EffectParameterSlot::slotValueChanged(double v) {
+                XmlParse::selectNodeQString(parameterElement,
+                                            EffectXml::ParameterLinkType))));
+        m_pControlLinkInverse->set(
+            XmlParse::selectNodeDouble(parameterElement,
```
This and the calls above can fail. I think it should go to default in this case, right?

Be-ing commented:
> the Super and Meta should either be unchanged and overwrite the stored parameter values, to keep the controller in sync (A)

It is up to the controller mapping to query the controller for the position of knobs upon mapping initialization. If the metaknobs or superknob were out of sync with parameters when closing Mixxx, the new values indicating the actual position of the controller knobs will override that.

daschuer commented on src/effects/effectchainslot.cpp:
```diff
@@ -431,3 +433,64 @@ void EffectChainSlot::slotChannelStatusChanged(const QString& group) {
+    return chainElement;
+}
+
+void EffectChainSlot::loadValuesFromXml(const QDomElement& effectChainElement) {
```
loadEffectsFromXml

daschuer commented on src/effects/effectbuttonparameterslot.cpp:
```diff
@@ -107,3 +109,33 @@ void EffectButtonParameterSlot::slotValueChanged(double v) {
+        m_pControlValue->reset();
+    } else {
+        m_pControlValue->set(
+            XmlParse::selectNodeDouble(parameterElement,
```
This can fail, adopt the default.

daschuer commented on src/effects/effectbuttonparameterslot.cpp:
```diff
@@ -107,3 +109,33 @@ void EffectButtonParameterSlot::slotValueChanged(double v) {
     }
 }
+
+QDomElement EffectButtonParameterSlot::toXML(QDomDocument* doc) const {
```
The toXML and loadValuesFromXml calls are not symetric. 
I would prefere to handle EffectXml::ParameterId, in EffectSlot::toXML

daschuer commented:
> I don't understand why we would store the parameter values but not the metaknob and superknob values. It's possible the user intentionally set parameters out of sync with the metaknob and superknob. I don't think Mixxx should automatically override that.

This XML format might be used to load and store effects during runtime. Than switching stored effect presets should follow the same rules as loading an default effect. 

What is the use case for storing out of sync values? How will the user get into this state intentionally?  
   


>> the Super and Meta should either be unchanged and overwrite the stored parameter values, to keep the controller in sync (A)

> It is up to the controller mapping to query the controller for the position of knobs upon mapping initialization. 

This is a feature that only some controllers support. 

> If the metaknobs or superknob were out of sync with parameters when closing Mixxx, the new values indicating the actual position of the controller knobs will override that.

I do not think this can be done reliably throwout all controller models with different focusing features on the first load. If you consider to load a effect preset on the fly this will fail always.

Be-ing commented:
> This XML format might be used to load and store effects during runtime. Than switching stored effect presets should follow the same rules as loading an default effect.

I'm not sure it should follow the same rules. The purpose of this PR is to save the state of effects as they were when Mixxx shuts down and restore them to that state on startup. I am doubtful there should be any caveats to that.

rryan commented:
This PR is only about loading and storing at startup/shutdown only -- I think fully saving the state of the effect units makes sense (to that end, why not dry/wet and the effect enable switches? I guess group enable state is saved via CO currently, i'd support moving it into the XML).

daschuer commented:
OK, than lets focus on the startup and shutdown issues. 

I am not sure about the enable switches. Could it be a problem if the an effect is left nearly silenced and the user can't figure out how to fix this issue the following week, because effects are out of sight? 

I have also not understand what is the benefit to store an out of sync situation. 
I understand @Be-ing idea that In case of a restart of Mixxx there should be as much as possible stay in sync. Unfortunately the controls will be probably out of sync anyway. If we want to fully restore the controller state without any out of sync issue, It is required to store the the focus state instead. That would allow to adopt the default of the effect that was controlled before.

Be-ing commented:
> I am not sure about the enable switches. Could it be a problem if the an effect is left nearly silenced and the user can't figure out how to fix this issue the following week, because effects are out of sight?

Yes, but that is much less of an issue than the situation I described in #1103.

> why not dry/wet and the effect enable switches? I guess group enable state is saved via CO currently, i'd support moving it into the XML

Effect enable switches should always be off when Mixxx starts as discussed in #1103. I'm not sure about the dry/wet knob.

> I have also not understand what is the benefit to store an out of sync situation.

The benefit is that the effects will sound exactly as they did the last time Mixxx ran.

> I understand @Be-ing idea that In case of a restart of Mixxx there should be as much as possible stay in sync. Unfortunately the controls will be probably out of sync anyway. If we want to fully restore the controller state without any out of sync issue, It is required to store the the focus state instead. That would allow to adopt the default of the effect that was controlled before.

What will be out of sync? The controller and the saved effects? If they're out of sync with the controller, the only remedy for that is querying the controller for the current position of the knobs. The focus state does persist across restarts. If you mean the effect parameters will be out of sync with the metaknobs, they will only be that way if the user set them like that before closing Mixxx before.

daschuer commented:
>> I have also not understand what is the benefit to store an out of sync situation.

> The benefit is that the effects will sound exactly as they did the last time Mixxx ran.

That is my goal as well. But there is no need that the meta knob stays out of sync. If Instead the controller mapping goes out of sync, the result would be the same, for the controller user, but the GUI user starts always in sync. 

What will be out of sync? The controller and the saved effects? 

If you for example switch between Mataknob and Parameter focus. The controller knobs are out of sync. So it depends on the focus last focus setting, if the restart was a switch, and the stored mataknob value matches with the knobs, or not.

> If they're out of sync with the controller, the only remedy for that is querying the controller for the current position of the knobs. The focus state does persist across restarts. 

How does this work? Does the controller store this in its persistent memory? 

A common use case is also that the controller was just installed, where it is unikely that the knob have not moved during transport.

rryan commented:
> But there is no need that the meta knob stays out of sync.

This presumes there is a position of the metaknob that is in sync. 

What I don't understand is how we should handle the case where there is an unsatisfiable set of constraints. Picking the first linked parameter to set the metaknob is simply going to put another linked parameter out of sync in this situation -- so it seems no better to me than doing nothing at all. 

Why do we need to fix things at all in this situation -- this is why we have soft takeover, right?

daschuer commented:
>>  But there is no need that the meta knob stays out of sync.

> This presumes there is a position of the metaknob that is in sync.

That is a follow up from our discussion in https://github.com/mixxxdj/mixxx/pull/1148 
If we decide It is a good way to adopt the first linked parameter there, suites here as well.

For my RMX 2 a metaknob that adopts the mapped parameter knob value would be ideal since I have an endless super knob. But there are also other ideas around to solve this.

Be-ing commented:
>  But there is no need that the meta knob stays out of sync.

Good point. 

> What I don't understand is how we should handle the case where there is an unsatisfiable set of constraints. Picking the first linked parameter to set the metaknob is simply going to put another linked parameter out of sync in this situation -- so it seems no better to me than doing nothing at all.

Also a good point.

> How does this work? Does the controller store this in its persistent memory?

No, focused_effect and show_parameters are persistent ControlObjects.

daschuer commented:
> No, focused_effect, show_parameters, and show_focus are persistent ControlObjects.

Ah, OK, I did not notice that. So it actually makes sense to store the metaknob and super knob values. 

I think we should store the focused_effect, show_parameters, and show_focus togehter with super and meta, since they are only valid together. 

I think it is better to store them as persistent controls, that would allow us later to reuse the xml format to store effect presets.

Be-ing commented:
> I think it is better to store them as persistent controls, that would allow us later to reuse the xml format to store effect presets.

Yes, likewise with the group enable states.

Be-ing committed:
avoid reusing loadValuesFromXml method name for slot classes at different levels

Be-ing committed:
small refactor to make toXML/loadEffectSlotToXML symmetrical

Be-ing committed:
check for valid value in effect XML file, otherwise keep default

Be-ing committed:
toXML -> toXml

Be-ing committed:
support default values for persistent ControlPotmeters

Be-ing committed:
make dry/wet knob persistent

Be-ing commented:
I think this PR is ready for merge. I don't think Mixxx should be trying to overcompensate for a user's potential misunderstanding. That could be even more confusing when a user, for whatever reason, intentionally sets the parameters and metaknob to an out of sync state. IMO restoring exactly the state the user had when they shut down Mixxx is the most straightforward behavior.

The group enable switches and dry/wet knob are persistent COs instead of stored in the XML because I don't think they would make sense to store for exporting & importing chains.

Be-ing commented:
@rryan ready for merge?

JosepMaJAZ commented:
From appveyor's log, it seems that there is a bug that prevents it to compile on Windows:

src\effects\effectchainslot.cpp(452): error C2440: 'static_cast': cannot convert from 'double' to 'EffectChain::InsertionType'
src\effects\effectchainslot.cpp(452): note: Conversions between enumeration and floating point values are no longer allowed
==============================

+    XmlParse::addElement(*doc, chainElement, EffectXml::ChainInsertionType,
 +                         EffectChain::insertionTypeToString(
 +                           static_cast<EffectChain::InsertionType>(
 +                              m_pControlChainInsertionType->get())));

Be-ing commented:
Ack, thanks for pointing that out.

Be-ing committed:
fix Windows build

explicitly cast double to int before casting to enum class

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into effect_xml

Be-ing commented:
Now AppVeyor is failing for an [unrelated reason](https://ci.appveyor.com/project/mixxxdj/mixxx/build/master-1447/job/3lv9x699k6jw670m#L6980). @sblaisot, do you know what's going on?

Be-ing commented:
Can someone test compiling this on Windows to check if I actually fixed the build error? @JosepMaJAZ?

JosepMaJAZ commented:
There needs to be two more casts, in line 128 and 268. The rest seems to compile fine.

Be-ing commented:
Thank you for testing. What file are those lines in? src/effects/effectchainslot.cpp does not seem to be the file you are referring to.

JosepMaJAZ commented:
diff --git a/src/effects/effectparameterslot.cpp b/src/effects/effectparameterslot.cpp
index 0abbb53..0e5344d 100644
--- a/src/effects/effectparameterslot.cpp
+++ b/src/effects/effectparameterslot.cpp
@@ -125,7 +125,8 @@ void EffectParameterSlot::slotParameterValueChanged(double value) {
 void EffectParameterSlot::slotLinkTypeChanging(double v) {
     m_pSoftTakeover->ignoreNext();
     EffectManifestParameter::LinkType newType =
-        static_cast<EffectManifestParameter::LinkType>(v);
+        static_cast<EffectManifestParameter::LinkType>(
+            static_cast<int>(v));
     if (newType == EffectManifestParameter::LinkType::LINKED_LEFT ||
         newType == EffectManifestParameter::LinkType::LINKED_RIGHT ||
         newType == EffectManifestParameter::LinkType::LINKED_LEFT_RIGHT) {
@@ -265,7 +266,7 @@ QDomElement EffectParameterSlot::toXml(QDomDocument* doc) const {
                              EffectXml::ParameterLinkType,
                              EffectManifestParameter::LinkTypeToString(
                                 static_cast<EffectManifestParameter::LinkType>(
-                                    m_pControlLinkType->get())));
+                                    static_cast<int>(m_pControlLinkType->get()))));
         XmlParse::addElement(*doc, parameterElement,
                              EffectXml::ParameterLinkInversion,
                              QString::number(m_pControlLinkInverse->get()));

Be-ing committed:
fix Windows build

thanks @JosepMaJAZ

Be-ing commented:
Okay, builds are working on each OS now. AppVeyor says it is failing because of the issue fixed in #1232. There seems to be a bug in the Travis scripts because that same test is failing but Travis is reporting that it was successful.

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into effect_xml

Be-ing commented:
Yay AppVeyor build works now :)

Be-ing commented:
ping @rryan: Is there anything left to do here? Ready for merge?

rryan commented on src/effects/effectslot.cpp:
```diff
@@ -268,3 +272,62 @@ void EffectSlot::slotEffectMetaParameter(double v, bool force) {
+
+    QDomElement parametersElement = doc->createElement(EffectXml::ParametersRoot);
+
+    for (const auto& pParameter : m_parametersById) {
```
Since m_parametersById is only used in serializing/deserializing, I wonder if it makes sense to build it on the fly instead of storing at the class level (which requires keeping it in sync as new effects are loaded). 

Also, why a QMap? did you want to the items to be sorted by ID in the XML (instead of being in the parameter order in the manifest).

Be-ing replied:
> Since m_parametersById is only used in serializing/deserializing, I wonder if it makes sense to build it on the fly instead of storing at the class level (which requires keeping it in sync as new effects are loaded).

I don't think so. EffectSlot::loadEffectSlotFromXml needs to check that the parameter IDs in the XML match IDs in the effect's manifest. m_parametersById is built in EffectSlot::loadEffect when the Effect is loaded is loaded from the XML, which happens before EffectSlot::loadEffectSlotFromXml is called. (EffectSlot::loadEffect is called by EffectChainSlot::slotChainEffectChanged when the EffectChain emits its effectChanged signal in EffectChain::addEffect, which is called from EffectChain::createFromXml).

> Also, why a QMap? did you want the items to be sorted by ID in the XML (instead of being in the parameter order in the manifest)?

I used a QMap to pair the EffectParameterSlotBasePointers with the corresponding manifests' ID (which is a QString). EffectSlot::loadEffectSlotFromXml needs to lookup an EffectParameterSlotBasePointer from the ID string in the XML. It would be a good idea to serialize them in order when writing the XML though. Then when support for users rearranging parameters is implemented that could be represented in the XML file with the order of the `<Parameter>` elements.

Be-ing replied:
Hmmm... I'll refactor this a bit to preserve parameter order.

rryan commented on src/effects/effectslot.cpp:
```diff
@@ -268,3 +272,62 @@ void EffectSlot::slotEffectMetaParameter(double v, bool force) {
+}
+
+void EffectSlot::loadEffectSlotFromXml(const QDomElement& effectElement) {
+    if (effectElement.text().isEmpty()) {
```
`QDomElement::text()` requires recursively building up a string of all the children's `.text()`. I think [QDomNode::hasChildNodes](https://doc.qt.io/qt-4.8/qdomnode.html#hasChildNodes) does what you want here but will be far cheaper?

https://code.qt.io/cgit/qt/qt.git/tree/src/xml/dom/qdom.cpp#n4565

rryan commented on src/effects/effectslot.cpp:
```diff
@@ -268,3 +272,62 @@ void EffectSlot::slotEffectMetaParameter(double v, bool force) {
+        if (parameterNode.isElement()) {
+            const QString id = XmlParse::selectNodeQString(parameterNode,
+                                                           EffectXml::ParameterId);
+            if (m_parametersById.contains(id)) {
```
This pattern does two O(log n) lookups in the QMap, you can avoid it like this:

```c++
EffectParameterSlotBasePointer pParameterSlot = m_parametersById.value(id);
if (pParameterSlot) { 
}
```
since QMap::value returns a default-constructed value if it's not present

rryan commented on src/effects/effect.cpp:
```diff
@@ -172,32 +173,15 @@ EffectParameter* Effect::getButtonParameterForSlot(unsigned int slotNumber) {
-    QString effectId = XmlParse::selectNodeQString(element, "Id");
+    // Empty <Effect/> elements are used to preserve chain order
+    // when there are empty slots at the beginning of the chain.
+    if (element.text().isEmpty()) {
```
`!element.hasChildNodes()` instead?

(see my other comment)

rryan commented on src/effects/effectbuttonparameterslot.cpp:
```diff
@@ -107,3 +109,33 @@ void EffectButtonParameterSlot::slotValueChanged(double v) {
+    if (m_pEffectParameter == nullptr) {
+        return;
+    }
+    if (parameterElement.text().isEmpty()) {
```
`!element.hasChildNodes()` instead?

(see my other comment)

rryan commented on src/effects/effectchainslot.cpp:
```diff
@@ -430,3 +432,64 @@ void EffectChainSlot::slotChannelStatusChanged(const QString& group) {
+}
+
+void EffectChainSlot::loadChainSlotFromXml(const QDomElement& effectChainElement) {
+    if (effectChainElement.text().isEmpty()) {
```
`!element.hasChildNodes()` instead?

(see my other comment)

rryan commented on src/effects/effectparameterslot.cpp:
```diff
@@ -242,3 +254,62 @@ void EffectParameterSlot::slotValueChanged(double v) {
+    if (m_pEffectParameter == nullptr) {
+        return;
+    }
+    if (parameterElement.text().isEmpty()) {
```
`!element.hasChildNodes()` instead?

(see my other comment)

rryan commented on src/effects/effectslot.cpp:
```diff
@@ -268,3 +272,62 @@ void EffectSlot::slotEffectMetaParameter(double v, bool force) {
+
+    QDomElement parametersElement = XmlParse::selectElement(effectElement,
+                                                            EffectXml::ParametersRoot);
+    if (parametersElement.text().isEmpty()) {
```
`!element.hasChildNodes()` instead?

(see my other comment)

rryan commented on src/effects/effectslot.cpp:
```diff
@@ -268,3 +272,62 @@ void EffectSlot::slotEffectMetaParameter(double v, bool force) {
+    if (effectElement.text().isEmpty()) {
+        return;
+    }
+
```
Since loading the effect chain into the effect chain slot and loading the chain slot's XML config happen separately, want to add a sanity check here that the XML effect ID matches the loaded effect ID and print a warning if it doesn't?

rryan commented on src/effects/effectslot.cpp:
```diff
@@ -173,12 +175,14 @@ void EffectSlot::loadEffect(EffectPointer pEffect) {
             addEffectButtonParameterSlot();
         }
 
+        m_parametersById.clear();
```
Please clear `m_parametersById` in `EffectSlot::clear()` as well

rryan reviewed (approved):
I did one more careful review and found a few minor things to fix. LGTM, nice work!

Be-ing committed:
Merge remote-tracking branch 'upstream/master' into effect_xml

Be-ing commented on src/effects/effectparameterslot.cpp:
```diff
@@ -2,35 +2,40 @@
     QString itemPrefix = formatItemPrefix(iParameterSlotNumber);
     m_pControlLoaded = new ControlObject(
             ConfigKey(m_group, itemPrefix + QString("_loaded")));
+    m_pControlType = new ControlObject(
```
They are deleted in the parent EffectParameterSlotBase destructor. Deleting them here causes a segfault. I'll add a comment.

rryan replied:
Oh oops, that's not confusing at all :) -- thanks.

Be-ing committed:
add missing function parameter

Be-ing committed:
clarifying comment

Be-ing committed:
organize EffectParameterSlot constructor & destructor

Be-ing committed:
efficiently check for empty XML elements

QDomNode::text recursively gets the text for each child node

Be-ing committed:
keep effect parameters in order in XML file

In the future, it should be possible to rearrange effect parameters and
store that in XML with the order of the <Parameter> elements.

rryan commented:
Just did a final test and everything seems to work well. For fun, I changed Flanger's effect ID to the unicode poop symbol and I couldn't get it to load in an effect unit so I didn't get to test whether the UTF-8 encoded poop actually ended up in the XML file :P. I settled for adding accented characters to the EffectChain's description and verifying they were preserved across restarts. LGTM, thanks!

Be-ing commented:
Thanks for review and testing! 2.1 will have a much better user experience for effects. :)

> unicode poop symbol

I did not know such a thing existed. o.O

daschuer commented:
In current Master, I cannot use prev_effect or prev_chain in Shade / LateNight.
Is this due to this PR?

daschuer commented on src/effects/effectsmanager.cpp:
```diff
@@ -218,15 +218,15 @@ EffectRackPointer EffectsManager::getEffectRack(const QString& group) {
-    pStandardRack->addEffectChainSlot();
-    pStandardRack->addEffectChainSlot();
-    pStandardRack->addEffectChainSlot();
-    pStandardRack->addEffectChainSlot();
```
I think the regression is caused by the missing empty Slots in case of no stored effect are available.

daschuer commented on src/effects/effectsmanager.cpp:
```diff
@@ -218,15 +218,15 @@ EffectRackPointer EffectsManager::getEffectRack(const QString& group) {
-    pStandardRack->addEffectChainSlot();
-    pStandardRack->addEffectChainSlot();
+    for (auto chain : savedChains) {
+        pStandardRack->addEffectChainSlot(chain.first, chain.second);
```
Please use here named members instead pointless "first" and "second"

Be-ing replied:
This is used only once. I think creating a struct just for this one use would be more trouble than its worth.

daschuer replied:
It already causes trouble and it tends to spread out later.

Be-ing commented:
> In current Master, I cannot use prev_effect or prev_chain in Shade / LateNight.

next_chain is working. It looks broken at first because there are 4 chains created that do not have any effects loaded and LateNight has no way to load different effects in a chain yet. Something is wrong with prev_chain though. prev_chain works for the chains made with the premade chain hack, which should be removed as soon as Shade and LateNight are updated, but it isn't working for the apparently blank chains.

next/prev_effect are working fine in Shade.

daschuer commented:
It looks like we have a conceptual issue here. 

The content of the xml file also creates the COs of the effects. This breaks the original concept, that the COs are independent from the loaded effect chains. 

Mixxx, or the loaded skin should set up the Effect slots and the stored chains from the XML is just responsible for the filling, loading the stored effects. 
If there are ore or less stored effects in the XML file, it should be handled gracefully.

daschuer commented:
> next/prev_effect are working fine in Shade.

No, sorry. Not if there is no XML file.

Be-ing commented:
> No, sorry. Not if there is no XML file.

I cannot reproduce this.

daschuer commented:
``` 
   for (auto chain : savedChains) {
        pStandardRack->addEffectChainSlot(chain.first, chain.second);
    }
```
Is skipped if there is no saved chain. And this instantiates and connects the prev_effect and next_effect COs. Without a connection nothing happens. 

But I think that the COs are not connected before is the real issue.

Be-ing commented:
> Is skipped if there is no saved chain.

No it is not. EffectChainManager::loadEffectChains will return 4 empty EffectChainPointer, QDomElement pairs if the file cannot be read or parsed.

daschuer commented:
For some reason I have found this file in my .mixxx folder 

```xml
<?xml version='1.0' encoding='utf-8'?>
<MixxxEffects schemaVersion="0">
 <Rack>
  <Group>[EffectRack1]</Group>
  <Chains/>
 </Rack>
</MixxxEffects>
```

If I delete it, the issue is gone. What could be the reason for getting it?

daschuer commented:
>> EffectChainManager::loadEffectChains will return 4 empty EffectChainPointer, QDomElement pairs if the file cannot be read or parsed.

IMHO this should not be in responds to loadEffectChains. Mixxx should setup the Chains before and load what is in the file. This would make the feature more robust against user edits.

Be-ing commented:
> If I delete it, the issue is gone. What could be the reason for getting it?

That was probably from testing an early version of this PR. Users should never have such a file.

The feature could be more robust against unusual XML files by ensuring that EffectChainManager::loadEffectChain always returns 4 EffectChainPointer, QDomElement pairs. The XML file you posted is a valid XML file, so it looked for effect chains to load, but ended up returning none (which is not the same as returning null EffectChainPointers).
